ionic-team/stencil
# feat(compiler): generate component custom event types with HTML target

## Pull request checklist

Please check if your PR fulfills the following requirements:

## Pull request type

 

 

Please check the type of change your PR introduces:


## What is the current behavior?


The generated types for a web components custom event, does not include type definitions for the `target` value.

GitHub Issue Number: #3291


## What is the new behavior?


- Stencil compiler will generate new exported interfaces in `components.d.ts` for all web component that have custom events.
- The generated interfaces includes the typing to the generated HTML element type for the event target and accepts a generic for the custom event detail.
- Stencil compiler will generate the web component custom event type with the new type, allowing usage with the custom event and the generated HTML element type for the `target`. 

## Does this introduce a breaking change?




## Testing



The steps outlined below can apply to any Stencil project, but will outline against Ionic Framework:
1. Build and pack a local version of this branch
2. Install the .tgz into Ionic Framework `/core` project.
3. Run `npm run build` to build Stencil
4. See that in `components.d.ts` that there is a new namespace called `ComponentEvents`.
5. Build the react package, by running `npm run build` from `packages/react`
6. Sync the build changes with the react test app, by running `npm run sync` from `packages/react/test-app`.
7. Run the test-app `npm run start`
8. Modify any of these `.tsx` files (i.e. App.tsx) to use a web component with a JSX event listener that accesses a custom public method on the web component:
```tsx
<IonInfiniteScroll onIonInfinite={ev => ev.target.complete()}></IonInfiniteScroll>
```
9. Validate that there is no typing errors/exceptions. Intellisense/tsc is able to determine the type of the `complete()` function from `infinite-scroll.tsx`. 

## Other information



Closing #3286 in favor of this PR.
